library/nist_kw.c

```diff
@@ -0,0 +1,649 @@
+    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
+    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
+} while( 0 )
+#endif
+
+/*
+ * Initialize context
+ */
+void mbedtls_nist_kw_init( mbedtls_nist_kw_context *ctx )
+{
+    memset( ctx, 0, sizeof( mbedtls_nist_kw_context ) );
+}
+
+int mbedtls_nist_kw_setkey( mbedtls_nist_kw_context *ctx,
+                            mbedtls_cipher_id_t cipher,
+                            const unsigned char *key,
+                            unsigned int keybits,
+                            const int is_wrap )
+{
+    int ret;
+    const mbedtls_cipher_info_t *cipher_info;
+
+    /*
+     * SP 800-38F currently defines AES cipher as the only block cipher allowed, but:
+     * "For KW and KWP, the underlying block cipher shall be approved, and the block size shall be
+     *  128 bits. Currently, the AES block cipher, with key lengths of 128, 192, or 256 bits, is the only
+     *  block cipher that fits this profile."
+     */
+    if( cipher != MBEDTLS_CIPHER_ID_AES )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cipher_info = mbedtls_cipher_info_from_values( cipher, keybits, MBEDTLS_MODE_ECB );
+    if( cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    if( cipher_info->block_size != 16 )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+
+    if( ( ret = mbedtls_cipher_setup( &ctx->cipher_ctx, cipher_info ) ) != 0 )
+        return( ret );
+
+    if( ( ret = mbedtls_cipher_setkey( &ctx->cipher_ctx, key, keybits,
+                               is_wrap ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT ) ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * KW-AE as defined in SP 800-38F section 6.2
+ * KWP-AE as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx, mbedtls_nist_kw_mode_t mode,
+                          const unsigned char *input, size_t in_len,
+                          unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t semiblocks = 0;
+    size_t s ;
+    size_t olen, i, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char* R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char* A = output ;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+    /*
+     * Generate the String to work on
+     */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the plaintext length for KW
+         * must be between 2 to 2^54-1 semiblocks inclusive.
+         */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
```

Comment:
What does `UINT_MAX` have to do with this check? Change to `#if SIZE_MAX > 0x1FFFFFFFFFFFFF8`.

Reply:
@RonEld: I am not sure what are the details of the discussion in this PR, but I do not think we should change the checks in pk_wrap and pk because the problem there is that at some point we have to cast a `size_t` to a `unsigned int` so UINT_MAX does make sense there as it is the maximum value that an `unsigned int` can hold. But in this PR you dont seem to be using `unsigned int`, so thats probably why @gilles-peskine-arm pointed out this bit of code.

Reply:
The objective here is not to know whether `size_t` is less than `0x1FFFFFFFFFFFFF8`.
The objective here is to avoid compilation warnings of type mismatch comparison, and to confirm the `size_t` is a 64 bit type. If it's a 32 bit type, there is no point in the runtime comparison. Do really see `size_t` as being a 48 bit size?
If it will make you happy, I'll change the comparison value.

Reply:
The objective is to avoid a compilation warning which is triggered when `0x1FFFFFFFFFFFFF8` is outside the range of `size_t`. It's unlikely that the size of `size_t` would be anything but 32 or 64 (or 16 but we don't support that). However, by assuming that these are the only possible sizes, you're making the code harder to understand (its correctness depends on an unstated assumption about the platform) and less portable, and you gain absolutely nothing.